adapters/openrtb_util.go

```diff
@@ -3,35 +3,136 @@ package adapters
 import (
 	"github.com/prebid/prebid-server/pbs"
 
+	"errors"
 	"github.com/prebid/openrtb"
 )
 
-func makeOpenRTBGeneric(req *pbs.PBSRequest, bidder *pbs.PBSBidder, bidderFamily string) openrtb.BidRequest {
+func min(x, y int) int {
+	if x < y {
+		return x
+	}
+	return y
+}
+
+func mediaTypeInSlice(t pbs.MediaType, list []pbs.MediaType) bool {
+	for _, b := range list {
+		if b == t {
+			return true
+		}
+	}
+	return false
+}
+
+func commonMediaTypes(l1 []pbs.MediaType, l2 []pbs.MediaType) []pbs.MediaType {
+	res := make([]pbs.MediaType, min(len(l1), len(l2)))
+	i := 0
+	for _, b := range l1 {
+		if mediaTypeInSlice(b, l2) {
+			res[i] = b
+			i = i + 1
+		}
+	}
+	return res[:i]
+}
+
+func makeBanner(unit pbs.PBSAdUnit) *openrtb.Banner {
+	return &openrtb.Banner{
+		W:        unit.Sizes[0].W,
+		H:        unit.Sizes[0].H,
+		Format:   unit.Sizes,
+		TopFrame: unit.TopFrame,
+	}
+}
 
-	imps := make([]openrtb.Imp, len(bidder.AdUnits))
-	for i, unit := range bidder.AdUnits {
+func makeVideo(unit pbs.PBSAdUnit) *openrtb.Video {
+	// empty mimes array is a sign of uninitialized Video object
+	if len(unit.Video.Mimes) < 1 {
+		return nil
+	}
+	mimes := make([]string, len(unit.Video.Mimes))
+	copy(mimes, unit.Video.Mimes)
+	pbm := make([]int8, 1)
+	pbm[0] = unit.Video.PlaybackMethod
+	return &openrtb.Video{
+		MIMEs:          mimes,
+		MinDuration:    unit.Video.Minduration,
+		MaxDuration:    unit.Video.Maxduration,
+		W:              unit.Sizes[0].W,
+		H:              unit.Sizes[0].H,
+		StartDelay:     unit.Video.Startdelay,
+		PlaybackMethod: pbm,
+	}
+}
+
+func makeOpenRTBGeneric(req *pbs.PBSRequest, bidder *pbs.PBSBidder, bidderFamily string, allowedMediatypes []pbs.MediaType, singleMediaTypeImp bool) (openrtb.BidRequest, error) {
+
+	imps := make([]openrtb.Imp, len(bidder.AdUnits)*len(allowedMediatypes))
+	ind := 0
+	impsPresent := false
+	for _, unit := range bidder.AdUnits {
 		if len(unit.Sizes) <= 0 {
+			ind = ind + 1
+			continue
+		}
+		unitMediaTypes := commonMediaTypes(unit.MediaTypes, allowedMediatypes)
+		if len(unitMediaTypes) == 0 {
 			continue
 		}
 
-		imps[i] = openrtb.Imp{
-			ID: unit.Code,
-			Banner: &openrtb.Banner{
-				W:        unit.Sizes[0].W,
-				H:        unit.Sizes[0].H,
-				Format:   unit.Sizes,
-				TopFrame: unit.TopFrame,
-			},
-			Secure: req.Secure,
-			// pmp
-			// ext
+		if singleMediaTypeImp {
+			for _, mType := range unitMediaTypes {
+				newImp := openrtb.Imp{
+					ID:     unit.Code,
+					Secure: req.Secure,
+				}
+				switch mType {
+				case pbs.MEDIA_TYPE_BANNER:
+					newImp.Banner = makeBanner(unit)
+				case pbs.MEDIA_TYPE_VIDEO:
+					video := makeVideo(unit)
+					if video == nil {
+						return openrtb.BidRequest{}, errors.New("Invalid AdUnit: VIDEO media type with no video data")
```

Comment:
Does a totally empty BidRequest have any valid uses?

If not, I think a `nil` here is more conventional. I'm not sure how strict of a Go convention it is... but all the code I've seen so far returns a value _or_ an error; I haven't seen any returning both yet.

Reply:
Since the return type here is a struct, Go does not allow to return `nil` here. I could either change the signature to return a pointer to struct, or return an empty struct. From what I understand, in Go it is customary to check on error first, and, if it is present, ignore the result. Therefore I decided to return an empty struct rather than change the signature to minimize change.

Reply:
TIL... ok, that makes sense. Didn't know Go worked that way.